### PR TITLE
Improve: new profiles conventional and ocamlformat

### PR DESCRIPTION
**Pull request by gpetiot**
See #656

- New `conventional` profile, with the following changes with respect to the default profile:
   * break-separators: from before to after-and-docked
   * break-cases: from fit to nested
   * break-infix-before-func: from true to false
   * break-sequences: from false to true
   * field-space: from tight to loose
   * sequence-style: from separator to terminator
- New `structure-left-margin` profile, equivalent to the default profile.

Didn't change the default values yet, so that people can try the `conventional` profile for the 0.9 release, and we can discuss about making the beginner values default for the next release.

### CODE REVIEW

**Comment by gpetiot** [comment]
Can (someone using this style) test the `beginner` on his/her codebase to see if its good enough? @samoht

**Comment by samoht** [comment]
I'm not a beginner :-)

**Comment by jberdine** [comment]
What about renaming `beginner` to something like `conventional`? My understanding is that it is intended to be as close to some currently popular style as the options we have allows, right?

**Comment by samoht** [comment]
I've run the `beginner` profile on irmin and the result looks good so far (need to check a bit more). The only strange choice is related to the formatting of records:

```
type t =
    { watches : (string, string -> unit Lwt.t) Hashtbl.t;
      files : (string, string) Hashtbl.t }

let t = {watches = Hashtbl.create 3; files = Hashtbl.create 13}
```

Would it be possible to have instead:
```
type t = {
    watches: (string, string -> unit Lwt.t) Hashtbl.t;
    files  : (string, string) Hashtbl.t;
  }

  let t = { watches = Hashtbl.create 3; files   = Hashtbl.create 13 }
```

Also, there is one [weird formatting](https://github.com/mirage/irmin/pull/582/files#diff-4846077446d291e891a3940001669978R135) of comments attached to record fields, but otherwise it looks pretty good :-)

I'll make it the default in my emacs settings and try to run it on save for a while.

**Comment by samoht** [comment]
Also I get a strange error message:
```
$ dune build @fmt
 ocamlformat src/irmin-mirage/.formatted/irmin_mirage.ml (exit 1)
(cd _build/default/src/irmin-mirage && /Users/thomas/git/irmin/_opam/bin/ocamlformat --impl irmin_mirage.ml --name ../../../../src/irmin-mirage/irmin_mirage.ml -o .formatted/irmin_mirage.ml)
ocamlformat: Cannot process "../../../../src/irmin-mirage/irmin_mirage.ml".
  Please report this bug at https://github.com/ocaml-ppx/ocamlformat/issues.
  BUG: comments dropped.
File "../../../../src/irmin-mirage/irmin_mirage.ml", line 437, characters 30-60:
Error: Comment (* no transaction is needed *) dropped.
```

That comment is at line 410, not 437:

```ocaml
          let repo = S.repo s.t in
          (* get the tree origin *)
          get_store_tree s >>= function
          | None -> f t >|= fun x -> Ok x  (* no transaction is needed *)
          | Some (origin, old_tree) ->
            let batch = { repo; tree = old_tree; origin } in
            let b = Batch batch in
```

**Comment by gpetiot** [comment]
> Also I get a strange error message:
> 
> ```
> $ dune build @fmt
>  ocamlformat src/irmin-mirage/.formatted/irmin_mirage.ml (exit 1)
> (cd _build/default/src/irmin-mirage && /Users/thomas/git/irmin/_opam/bin/ocamlformat --impl irmin_mirage.ml --name ../../../../src/irmin-mirage/irmin_mirage.ml -o .formatted/irmin_mirage.ml)
> ocamlformat: Cannot process "../../../../src/irmin-mirage/irmin_mirage.ml".
>   Please report this bug at https://github.com/ocaml-ppx/ocamlformat/issues.
>   BUG: comments dropped.
> File "../../../../src/irmin-mirage/irmin_mirage.ml", line 437, characters 30-60:
> Error: Comment (* no transaction is needed *) dropped.
> ```
> 
> That comment is at line 410, not 437:
> 
> ```ocaml
>           let repo = S.repo s.t in
>           (* get the tree origin *)
>           get_store_tree s >>= function
>           | None -> f t >|= fun x -> Ok x  (* no transaction is needed *)
>           | Some (origin, old_tree) ->
>             let batch = { repo; tree = old_tree; origin } in
>             let b = Batch batch in
> ```

Fixed in #668

**Comment by gpetiot** [comment]
> I've run the `beginner` profile on irmin and the result looks good so far (need to check a bit more). The only strange choice is related to the formatting of records:
> 
> ```ocaml
> type t =
>     { watches : (string, string -> unit Lwt.t) Hashtbl.t;
>       files : (string, string) Hashtbl.t }
> 
> let t = {watches = Hashtbl.create 3; files = Hashtbl.create 13}
> ```
> 
> Would it be possible to have instead:
> 
> ```ocaml
> type t = {
>     watches: (string, string -> unit Lwt.t) Hashtbl.t;
>     files  : (string, string) Hashtbl.t;
>   }
> 
> let t = { watches = Hashtbl.create 3; files = Hashtbl.create 13 }
> ```
> 
> Also, there is one [weird formatting](https://github.com/mirage/irmin/pull/582/files#diff-4846077446d291e891a3940001669978R135) of comments attached to record fields, but otherwise it looks pretty good :-)
> 
> I'll make it the default in my emacs settings and try to run it on save for a while.

Fixed in #669 (waiting for approval)

File: src/Conf.ml [hunk 1]
```diff
@@ -1414,10 +1414,10 @@ let (_profile : t option C.t) =
     [ ( "default"
       , Some default_profile
       , "$(b,default) sets each option to its default value." )
-    ; ( "beginner"
-      , Some beginner_profile
-      , "$(b,beginner) sets options for a beginner-friendly style, closer \
-         to imperative syntax." )
+    ; ( "conventional"
+      , Some conventional_profile
+      , "$(b,conventional) sets options for a closer to imperative syntax."
```
Comment (samoht) [inline src/Conf.ml]: missing word?

**Comment by samoht** [comment]
There are still missing spaces between `{`  and `}`:

```ocaml
let t = {watches = Hashtbl.create 3; files = Hashtbl.create 13}
```
instead of:
```ocaml
let t = { watches = Hashtbl.create 3; files = Hashtbl.create 13 }
```

Note very happy with the new default for pattern-matching clauses (and functions arguments maybe). For instance, this diff is not great: https://github.com/mirage/irmin/pull/582/files#diff-de0d3b7b9e3d49622c9910762e10a863L287. I'm not sure to understand what's wrong with:

```ocaml
match foo with
| Foo | Bar -> toto
| X -> toto
```

However, I can understand that having something like this would be (a bit?) confusing:

```ocaml
match foo with
| Toto | Titi | Xxxxxxx.... 
| Foo | Bar -> toto
| X -> toto
```

So I'm fine to add a break in this case ; but I don't think this should be done on every cases.

```ocaml
match foo with
| Toto | Titi | Xxxxxxx.... 
| Foo | Bar -> 
  toto
| X -> toto
```

**Comment by gpetiot** [comment]
I added `space-around-collection-expressions=true` for the spaces in the records.
I think `break-cases=toplevel` would be closer to the output you want, but I'm not sure it should be in this new profile (needs more testing).

**Comment by samoht** [comment]
Ok, `space-around-collection-expressions=true` looks good.

Now that irmin_mirage.ml can be parsed, I noticed an other strange format:

```ocaml
  module type KV_MAKER = functor
    (G: Irmin_git.G)
    (C: Irmin.Contents.S) ->
    S with type key = string list
       and type step = string
       and type contents = C.t
       and type branch = string
       and module Git = G
```
is formatted into:

```ocaml
  module type KV_MAKER = functor (G : Irmin_git.G) (C : Irmin.Contents.S) -> S
                                                                             with type
                                                                             key =
                                                                               string
                                                                               list
                                                                              and type
                                                                             step =
                                                                               string
                                                                              and type
                                                                             contents =
                                                                               C
                                                                               .t
                                                                              and type
                                                                             branch =
                                                                               string
                                                                              and module Git = G
```

**Comment by samoht** [comment]
And I'll try with `break-cases=toplevel`

**Comment by samoht** [comment]
So `break-cases=toplevel` looks closer to what I want indeed, but I've noticed that single-line pattern matching are broken into multiple lines? Would be great to not do this but I can live with that. 

```ocaml
-    let raw_value = function None -> None | Some v -> Some (raw_value v) in
+    let raw_value = function
+      | None -> None
+      | Some v -> Some (raw_value v)
+    in
```
Also, for is there any way to stack the arguments of anonymous functions as much as possible on the same lines? (e.g. in the irmin example with 260 arguments, that looks not very nice to put each one on their own line).

File: src/Conf.ml [hunk 2]
```diff
@@ -1305,6 +1305,18 @@ let default_profile =
...
 
+let conventional_profile =
+  { default_profile with
+    break_infix_before_func= false
+  ; break_separators= `After_and_docked
+  ; break_sequences= true
+  ; field_space= `Loose
+  ; indicate_nested_or_patterns= false
+  ; sequence_style= `Terminator
+  ; space_around_collection_expressions= true }
```
Comment (Julow) [inline src/Conf.ml]: This option does not work on type declaration combined with `break-separators=after-and-docked`:

```ocaml
type t = {x : int}
```

With `break-separators=after`:

```ocaml
type t = { x : int }
```

Comment (Julow) [reply src/Conf.ml]: https://github.com/ocaml-ppx/ocamlformat/pull/670

**Comment by gpetiot** [comment]
> Also, there is one [weird formatting](https://github.com/mirage/irmin/pull/582/files#diff-4846077446d291e891a3940001669978R135) of comments attached to record fields, but otherwise it looks pretty good :-)

@samoht I don't think we will fix that by the way, having comments before with 0-indent and comments after with 4-indent seems to work pretty well.
In the code you linked the comments are attached to the wrong fields (after the semicolon means they get attached to the next field), if you rewrite it like this:
```ocaml
type 'a t = {
  chunking : [`Max | `Best_fit];
  (* An handler to the underlying database. *)
  db : 'a CA.t;
  (* the size of chunks. *)
  chunk_size : int;
  (* the maximum number of children a node can have. *)
  max_children : int;
  (* the maximum length (in bytes) of data stored in one chunk *)
  max_data : int
}
```
each comment is correctly attached before the right field.

**Comment by gpetiot** [comment]
> Ok, `space-around-collection-expressions=true` looks good.
> 
> Now that irmin_mirage.ml can be parsed, I noticed an other strange format:
> 
> ```ocaml
>   module type KV_MAKER = functor
>     (G: Irmin_git.G)
>     (C: Irmin.Contents.S) ->
>     S with type key = string list
>        and type step = string
>        and type contents = C.t
>        and type branch = string
>        and module Git = G
> ```
> 
> is formatted into:
> 
> ```ocaml
>   module type KV_MAKER = functor (G : Irmin_git.G) (C : Irmin.Contents.S) -> S
>                                                                              with type
>                                                                              key =
>                                                                                string
>                                                                                list
>                                                                               and type
>                                                                              step =
>                                                                                string
>                                                                               and type
>                                                                              contents =
>                                                                                C
>                                                                                .t
>                                                                               and type
>                                                                              branch =
>                                                                                string
>                                                                               and module Git = G
> ```

I failed to fix this one (without having a lot of regressions) so far, so I created this issue #671. This issue shoudn't delay this PR anyway.

**Comment by jberdine** [comment]
I pushed a commit to this branch that does the renaming I suggested, expands the documentation of various options, and fixes up some accumulated non-alphabetized orderings. See what you think. Should be no functional change unless I missed something.

File: ocamlformat-help.txt [hunk 3]
```diff
@@ -297,13 +304,32 @@ OPTIONS
...
+           easier, so indentation or white space is avoided unless it helps
+           legibility; Attention has been given to making some syntactic
+           gotchas visually obvious. The compact profile is similar to
+           ocamlformat but opts for a generally more compact code style. The
+           sparse profile is similar to ocamlformat but opts for a generally
+           more sparse code style. The conventional profile aims to be as
+           familiar and "conventional" appearing as the available options
+           allow. The janestreet profile is used at Jane Street. The
+           "default" profile is ocamlformat with break-cases=fit. default is
+           deprecated and will be removed in version 0.10.
```
Comment (jberdine) [inline ocamlformat-help.txt]: It would be nice if there was a way to get more structure in this, but I don't see how to convince cmdliner to do so.

Comment (gpetiot) [reply ocamlformat-help.txt]: I didn't find a way to add some linebreaks.

**Comment by gpetiot** [comment]
Your last commit looks good to me, is it ready to be merged now?

**Pull request merged**